### PR TITLE
[rh:niquests] New network handler with HTTP2

### PR DESCRIPTION
**Pull request by Ousret**
Hello fellow maintainers,

This PR is probably going to be unexpected, but bear with me until the end :+1: 

Recently, _yt-dlp_ made the decision to support _Requests_ as one of the supported network adapters via https://github.com/yt-dlp/yt-dlp/pull/3668
For understandable reasons, the participants almost wanted to try _another client_ instead.

Today, I may make you even more satisfied with that choice. :rocket:  Let me introduce you to a replacement candidate. [Niquests](https://github.com/jawah/niquests). It is a drop-in replacement for the former that is no longer under feature freeze.

There's a ton of pros, but amongst them, you'll find:

- Native OS Trust Store, Certifi is gone!
- HTTP/2, and HTTP/3
- No more Python httplib for HTTP/1.1
- Faster ~10 to 20% without multiplexed, still non-negligible
- Can leverage a multiplexed connection to get up to 3X faster throughput
- Resilient to most bot detector
- Better security
- More reliable and compliant to standards

This current PR does not leverage the multiplexed aspect of HTTP/2, and HTTP/3 because it does require a bit more work.
I am willing to give it a shot if, you (_yt-dlp_ team) are open to it. A multiplexed connection is important to leverage the protocol's capabilities.

Now, you can have multiple concurrent requests without ever needing threads, multiple connections, etc... Potentially linked to https://github.com/yt-dlp/yt-dlp/issues/1918

If you are open to thinking about this, I will guarantee the utmost standard and protection against regression with _yt-dlp_. We will be in close contact to make sure everything is fine.

I truly think that we have a unique opportunity to bootstrap something cool.

Regards,

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [x] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: test/test_networking.py [hunk 1]
```diff
@@ -860,7 +860,7 @@ def request(self, *args, **kwargs):
         (lambda: urllib3.exceptions.DecodeError(), TransportError, None),
         (lambda: urllib3.exceptions.HTTPError(), TransportError, None),  # catch-all
         (
-            lambda: urllib3.exceptions.ProtocolError('error', http.client.IncompleteRead(partial=b'abc', expected=4)),
```
Comment (Ousret) [inline test/test_networking.py:863]: We no longer rely on the stdlib. That's why.

File: yt_dlp/networking/_requests.py [hunk 2]
```diff
@@ -276,16 +278,24 @@ def _check_extensions(self, extensions):
...
-        session.adapters.clear()
-        session.headers = requests.models.CaseInsensitiveDict({'Connection': 'keep-alive'})
-        session.mount('https://', http_adapter)
-        session.mount('http://', http_adapter)
+        certs = self._client_cert
+        session = RequestsSession(multiplexed=True, retries=False)
+
+        session.verify = self.verify
+
+        if "client_certificate_password" in certs and "client_certificate_key" in certs and certs["client_certificate_password"]:
```
Comment (Ousret) [inline yt_dlp/networking/_requests.py:286]: This block can be simplified.

File: yt_dlp/networking/_requests.py [hunk 3]
```diff
@@ -276,16 +278,24 @@ def _check_extensions(self, extensions):
...
 
     def _create_instance(self, cookiejar):
-        session = RequestsSession()
-        http_adapter = RequestsHTTPAdapter(
-            ssl_context=self._make_sslcontext(),
-            source_address=self.source_address,
-            max_retries=urllib3.util.retry.Retry(False),
-        )
-        session.adapters.clear()
-        session.headers = requests.models.CaseInsensitiveDict({'Connection': 'keep-alive'})
```
Comment (Ousret) [inline yt_dlp/networking/_requests.py:286]: It's automatically added (only when HTTP/1.1) in upstream deps.

File: yt_dlp/networking/_requests.py [hunk 4]
```diff
@@ -147,41 +131,59 @@ def read(self, amt: int = None):
...
-                 if isinstance(err, http.client.IncompleteRead)), None)
+                 if isinstance(err, urllib3.exceptions.IncompleteRead)), None)
             if ir_err is not None:
-                # `urllib3.exceptions.IncompleteRead` is subclass of `http.client.IncompleteRead`
-                # but uses an `int` for its `partial` property.
+                # `urllib3.exceptions.IncompleteRead` uses an `int` for its `partial` property.
                 partial = ir_err.partial if isinstance(ir_err.partial, int) else len(ir_err.partial)
                 raise IncompleteRead(partial=partial, expected=ir_err.expected) from e
+
+            # workaround until a tiny fix is delivered in urllib3-future.
```
Comment (Ousret) [inline yt_dlp/networking/_requests.py:140]: An edge case that is going to be addressed.

File: yt_dlp/networking/_requests.py (hunk 2 above)
Comment (Ousret) [reply yt_dlp/networking/_requests.py]: It is now simplified.

File: yt_dlp/networking/_requests.py (hunk 3 above)
Comment (Ousret) [reply yt_dlp/networking/_requests.py:286]: Was just a note for removal justification.

File: yt_dlp/networking/_requests.py (hunk 4 above)
Comment (Ousret) [reply yt_dlp/networking/_requests.py]: Addressed. Fixed.

File: test/test_networking.py (hunk 1 above)
Comment (Ousret) [reply test/test_networking.py:860]: Was just a note, a justification for the test parameter patch.

**Comment by Ousret** [comment]
FYI: I think the work done here can (potentially) supersede https://github.com/yt-dlp/yt-dlp/pull/7595 thus easing the maintenance burden.
On my side, the results are conclusive, maybe a run on anyone else can confirm or refute this statement.

The PR has been updated with (above) comments that are now resolved.
Feel free to ask anything about this work.

**Comment by coletdjnz** [comment]
Hi! Thanks for the PR. Your project seems very interesting, I'm keen to dig into it a bit when I get some time. Some of the features look very nice indeed :+1: .

Some background:

- I have had some questions as to why requests was added over more modern libraries. The quick answer is that it is stable, mature, widely used (and audited), and has very similar behaviour (with some tweaks) to be a "replacement" baseline for the aging urllib library we were using. There currently is no other candidate better than requests for this imo. So we need to keep requests as the default http/https handler.

- curl_cffi covers a very particular case and is only planned to be used in specific scenarios. It uses curl-impersonate to make the requests look *exactly* like they came from a real browser - from http layer down to the TLS layer (which you can't do with normal openssl python). This is to help curb fingerprinting in some cases (see https://github.com/orgs/yt-dlp/projects/6). https://github.com/yt-dlp/yt-dlp/pull/7595 also introduces an impersonation framework so we are not dependent on curl_cffi (or any other one library) to provide this support. 

A few things I notice of niquests from a brief look:
- looks to be very new (only a month or so old?) 
- may have/rely on non-Python components (can bring troubles for our build process - curl_cffi has this issue).  
- may rely on some non-production ready libraries (hface I was told is not production ready)

While these are not blockers for allowing us to support the library, we do need to take caution.

Given this, I would say at a minimum niquests will need to be it's own handler, and de-prioritized by default to start with (treated as experimental).  I am particularly interested in finding out what situations this can be useful in, as we can change handler priority per request :thinking:  (like we plan to do for curl_cffi).

We also do need some confidence that any library we add is going to be maintained and has some degree of code audit (either by us or the wider community). While our framework allows us to not depend on any one http library, we have very limited developer time so need to tread carefully.

There are some higher priority networking-related tasks we need to focus on first before this - such as a user facing/CLI method to change handler priority (which would allow this handler to be used). So be aware things may take some time (particularly with this networking stuff, from my experience :pensive:).

cheers

**Comment by pukkandan** [comment]
There are a few things we can start working on right now:
1. Separate requests and niquests into two files, creating any necessary abstraction to deal with common code. Until we implement a CLI option, you can set niquests to have priority in the PR so its features can be easily tested.
2. We are intentionally using certifi as opposed to system store and this should not be changed. Related code and dependencies should be removed
3. Test whether this can solve any of https://github.com/orgs/yt-dlp/projects/6 in place of curl_cffi. If it can bypass fingerprinting as efectively as curl, then we can prioritize this PR over it.

**Comment by coletdjnz** [comment]
> There are a few things we can start working on right now:
> 1. Separate requests and niquests into two files, creating any necessary abstraction to deal with common code. Until we implement a CLI option, you can set niquests to have priority in the PR so its features can be easily tested.
> 2. We are intentionally using certifi as opposed to system store and this should not be changed. Related code and dependencies should be removed
> 3. Test whether this can solve any of https://github.com/orgs/yt-dlp/projects/6 in place of curl_cffi. If it can bypass fingerprinting as effectively as curl, then we can prioritize this PR over it. If not, then the next biggest draw for us is HTTP2/3, but that's not nearly as high priority right now.

2: I think this is different than how Python and openssl work and what they read. been a while since I've looked into it though.

3: if this is not impersonating a browser but works then it would likely be because it's a new unique client that has yet to be fingerprinted and blocked. I don't see this as a replacement/higher priority than that.

**Comment by pukkandan** [comment]
> 2: I think this is different than how Python and openssl work and what they read. been a while since I've looked into it though.

We still have no reason to adopt it. certifi is perfect for our needs and provides a uniform store across devices and OSes as opposed to the inherent variability in any system store.

**Comment by pukkandan** [comment]
> 3: if this is not impersonating a browser but works then it would likely be because it's a new unique client that has yet to be fingerprinted and blocked. I don't see this as a replacement/higher priority than that, since it's only a matter of when it will stop working (similar to us changing our ciphers, TLS version, etc).

We still have quite a bit of work to get curl_cffi production ready last I checked. This PR is comparatively way smaller. I'm not saying we abandon your PR, just that it may be worth implementing this asap without waiting for it.

**Comment by Ousret** [comment]
Hello,

Awesome, I am glad that discussions emerged from this.

I am completely on board if you want to start progressively, let's agree that if you find that Niquests is both stable and resolves pains you may consider it your default.

I can, and will follow your guidance on implementation.

But first, let's tackle some of your legitimate raised concerns.

@coletdjnz 

Thank you for your initial review,

> I have had some questions as to why requests was added over more modern libraries. The quick answer is that it is stable, mature, widely used (and audited), and has very similar behaviour (with some tweaks) to be a "replacement" baseline for the aging urllib library we were using.

Interesting choice, and from what I saw in the PR, this is what I inferred, motivation speaking.

From a philosophical standpoint, I am curious that a project that is a fork, therefore less mature than the original, why wouldn't you recommend the older one? My guess is that this project is better. And it is actually better :+1: 

On a more concrete level, Niquests is well-tested and backward compatible, no tests have been stripped and I guess you saw that. So what Requests already guarantees is guaranteed by Niquests and more.
We are, as we speak, partnering up with Tidelift to get a higher level of insurance for the project. They reviewed it and accepted it!

> looks to be very new (only a month or so old?)

The package yes, but this adventure started way earlier, a year ago. 

> may have/rely on non-Python components

You are speaking of `qh3` which is optional. The required dependencies are pure Python.

> may rely on some non-production ready libraries (hface I was told is not production ready)
    
Fortunately, everything is production-ready. And already been tested in production in various companies that accepted to migrate. Still modest but shows great results.
"hface" is an idea, we just rewritten everything and tested it extensively. As you may have witnessed in our test suites.

> curl_cffi covers a very particular case and is only planned to be used in specific scenarios. It uses curl-impersonate to make the requests look exactly like they came from a real browser

There may be a bit of misconception on what actually happen. I could be mistaken, but they're nothing done there that Python cannot do. I can prove to you, on the lowest level, it looks nearly exactly as Firefox does it.
Niquests is the closest you can get without hazardous materials. I can get to 100% but at a cost of backward compatibility.
Which is not reasonable at the time.

Beware that "super"-cloud infrastructure cannot accurately ban this "fingerprint" as the difference starts to be too small, and the cost for them is unreasonable. Right now, they just ban excessive request consumption.

> We also do need some confidence that any library we add is going to be maintained and has some degree of code audit (either by us or the wider community). 

I hope that my contributions (other projects, like charset-normalizer) can give you a sense of my seriousness.

In addition to that, Tidelift will be active on the audit and security side.

> While our framework allows us to not depend on any one http library, we have very limited developer time so need to tread carefully.

I know that and have the utmost respect for it. 
Ask yourself, who of Requests or Niquests will listen and act on your needs today.

> If this is not impersonating a browser but works then it would likely be because it's a new unique client that has yet to be fingerprinted and blocked.

We will act on it when it shall happen. Don't worry. We have at least six to twelve months before that happen.

> I think this is different than how Python and openssl work and what they read. been a while since I've looked into it though.

Good catch. Yes, it goes deeper.

@pukkandan  

Thank you for your initial review

> We still have no reason to adopt it. certifi is perfect for our needs and provides a uniform store across devices and OSes as opposed to the inherent variability in any system store.

You are right. Uniformity may be a goal.
Unfortunately, there is a ton of situation where there is missing CAs, like in some countries, you may be required to pass in a specific proxy and the CA would only be available quickly through the system CA.
Or within any company that forces proxying everything, the CA will be in the system trust store.

If you want Certifi, Certifi it is. Of course.

**Comment by pukkandan** [comment]
> Unfortunately, there is a ton of situation where there is missing CAs, like in some countries, you may be required to pass in a specific proxy and the CA would only be available quickly through the system CA.
> Or within any company that forces proxying everything, the CA will be in the system trust store.

We already have a switch to use `context.load_default_certs` and nobody has ever asked for more. There is no point solving problems nobody cares about, and certainly not by adding a new dependency for it.

**Comment by Grub4K** [comment]
The reason we still require `certifi` is that we want to allow minimal dependencies. `requests`, `curl_cffi` and eventually `niquests` are supposed to all be optional dependencies. People should be able to select whatever dependencies they want, and it should still work. `yt-dlp` is not as opinionated in that regard.
The problem currently is that `requirements.txt` cannot reflect those things accurately. We would want to have a more granular `pyproject.toml` based dependency list. That's still in the works however.

## [`[rh:curlcffi]`](<https://github.com/yt-dlp/yt-dlp/pull/7595>)
Let me illustrate why `curl_cffi` has to be prioritized. Let's see what happens with cloudflare protected websites using `niquests`:

<details><summary>Kick: <code>yt-dlp -F https://kick.com/video/dd8c8125-cf6b-46c8-ad87-770507c389ec</code></summary>
```shell
[KickVOD] Setting up session
WARNING: [KickVOD] Unable to download webpage: HTTP Error 403: Forbidden
[KickVOD] Extracting URL: https://kick.com/video/dd8c8125-cf6b-46c8-ad87-770507c389ec
[KickVOD] dd8c8125-cf6b-46c8-ad87-770507c389ec: Downloading API JSON
ERROR: [KickVOD] dd8c8125-cf6b-46c8-ad87-770507c389ec: Unable to download JSON metadata: HTTP Error 403: Forbidden (caused by <HTTPError 403: Forbidden>); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
```
</details><details><summary>Crunchyroll: <code>yt-dlp -F --username user --pass pass https://www.crunchyroll.com/watch/G50UZXW05</code></summary>
```
[crunchyroll] Getting session id
[crunchyroll] Logging in
[crunchyroll] Extracting URL: https://www.crunchyroll.com/watch/G50UZXW05
[crunchyroll] Authenticating with grant_type=etp_rt_cookie
ERROR: [crunchyroll] G50UZXW05: Request blocked by Cloudflare; navigate to Crunchyroll in your browser, then pass the fresh cookies (with --cookies-from-browser or --cookies) and your browser's User-Agent (with --user-agent)
```
</details>

Both requests get blocked by cloudflare.
Here the same requests using `curl_cffi` with `--impersonate chrome` for comparison:

<details><summary>Kick: <code>yt-dlp -F https://kick.com/video/dd8c8125-cf6b-46c8-ad87-770507c389ec --impersonate chrome</code></summary>

```
[KickVOD] Setting up session
[KickVOD] Extracting URL: https://kick.com/video/dd8c8125-cf6b-46c8-ad87-770507c389ec
[KickVOD] dd8c8125-cf6b-46c8-ad87-770507c389ec: Downloading API JSON
[KickVOD] dd8c8125-cf6b-46c8-ad87-770507c389ec: Downloading m3u8 information
[info] Available formats for dd8c8125-cf6b-46c8-ad87-770507c389ec:
ID      EXT RESOLUTION FPS │  FILESIZE   TBR PROTO │ VCODEC      ACODEC
──────────────────────────────────────────────────────────────────────────
160p    mp4 284x160     30 │ ~ 1.73GiB  230k m3u8  │ avc1.4D401F mp4a.40.2
360p    mp4 640x360     30 │ ~ 4.74GiB  630k m3u8  │ avc1.4D401F mp4a.40.2
480p    mp4 852x480     30 │ ~12.50GiB 1663k m3u8  │ avc1.4D401F mp4a.40.2
720p60  mp4 1280x720    60 │ ~27.50GiB 3658k m3u8  │ avc1.4D401F mp4a.40.2
1080p60 mp4 1920x1080   60 │ ~54.81GiB 7291k m3u8  │ avc1.64002A mp4a.40.2
```
</details><details><summary>Crunchyroll: <code>yt-dlp -F --username user --pass pass https://www.crunchyroll.com/watch/G50UZXW05 --impersonate chrome</code></summary>

```
[crunchyroll] Getting session id
[crunchyroll] Logging in
[crunchyroll] Extracting URL: https://www.crunchyroll.com/watch/G50UZXW05
[crunchyroll] Authenticating with grant_type=etp_rt_cookie
[crunchyroll] G50UZXW05: Downloading object info JSON (cms)
[crunchyroll] Retrieving signed policy
[crunchyroll] G50UZXW05: Downloading stream info JSON (signed cms)
[crunchyroll] To get all formats of a hardsub language, use "--extractor-args crunchyrollbeta:hardsub=<language_code or all>". See https://github.com/yt-dlp/yt-dlp#crunchyrollbeta-crunchyroll for more info
[crunchyroll] G50UZXW05: Downloading adaptive_hls HLS manifest
[crunchyroll] G50UZXW05: Downloading chapter info
[info] Available formats for G50UZXW05:
ID                               EXT RESOLUTION FPS │   FILESIZE   TBR PROTO │ VCODEC      ACODEC    MORE INFO
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
adaptive_hls-hardsub-ar-SA-meta  mp4 multiple       │                  m3u8F │ unknown     unknown   [ja-JP] Quality selection URL
adaptive_hls-hardsub-de-DE-meta  mp4 multiple       │                  m3u8F │ unknown     unknown   [ja-JP] Quality selection URL
adaptive_hls-hardsub-en-US-meta  mp4 multiple       │                  m3u8F │ unknown     unknown   [ja-JP] Quality selection URL
adaptive_hls-hardsub-es-419-meta mp4 multiple       │                  m3u8F │ unknown     unknown   [ja-JP] Quality selection URL
adaptive_hls-hardsub-es-ES-meta  mp4 multiple       │                  m3u8F │ unknown     unknown   [ja-JP] Quality selection URL
adaptive_hls-hardsub-fr-FR-meta  mp4 multiple       │                  m3u8F │ unknown     unknown   [ja-JP] Quality selection URL
adaptive_hls-hardsub-it-IT-meta  mp4 multiple       │                  m3u8F │ unknown     unknown   [ja-JP] Quality selection URL
adaptive_hls-hardsub-pt-BR-meta  mp4 multiple       │                  m3u8F │ unknown     unknown   [ja-JP] Quality selection URL
adaptive_hls-hardsub-ru-RU-meta  mp4 multiple       │                  m3u8F │ unknown     unknown   [ja-JP] Quality selection URL
adaptive_hls-562                 mp4 428x240     24 │ ~ 97.55MiB  563k m3u8  │ avc1.42c015 mp4a.40.2 [ja-JP]
adaptive_hls-1093                mp4 640x360     24 │ ~189.62MiB 1094k m3u8  │ avc1.4d401e mp4a.40.2 [ja-JP]
adaptive_hls-2091                mp4 848x480     24 │ ~362.60MiB 2092k m3u8  │ avc1.4d401f mp4a.40.2 [ja-JP]
adaptive_hls-4108                mp4 1280x720    24 │ ~712.22MiB 4109k m3u8  │ avc1.640028 mp4a.40.2 [ja-JP]
adaptive_hls-8118                mp4 1920x1080   24 │ ~  1.37GiB 8119k m3u8  │ avc1.640028 mp4a.40.2 [ja-JP]
```
</details>

Both are not blocked and work. *This* is why `curl_cffi` is so important rn.

Having said that, h2 and h3 with a modern touch sounds pretty cool :+1:.

Adding it as an additional, optional handler would allow you to work on getting the above to work using the `niquests` framework. If the above could be made work, only then could `curl_cffi` be deprioritized. Hope you understand!

**Comment by Ousret** [comment]
Thank you for the quick reproducer. Yes, those sites are configured to have a pretty aggressive way to check the client.
It is still possible to mimic the actual browser without having to use curl impersonator. Especially when using HTTP/3 due to the higher control we have on the actual TLS handshake. Everything goes down to the cipher suite, client hello + extensions, and associated default headers.

Unfortunately, it is not Niquests highest priority right now, but I clearly noted this point. I will keep you notified in case.

> Having said that, h2 and h3 with a modern touch sounds pretty cool 👍.

:+1:  

I am going to modify that PR so that it is completely optional.
Also going to make sure the code base is not duplicated.

> The reason we still require certifi is that we want to allow minimal dependencies.

Would you be able to sustain this in the long term? Dependency-wise and maintenance burden-wise.

**Comment by Grub4K** [comment]
Just as a heads up, `pyproject.toml` (for optional groups) as well as #7595 are now merged so this can be considered again

**Comment by Ousret** [comment]
Great.
Since last we spoke, quite a few things evolved, and [Niquests](https://github.com/jawah/niquests) has improved.
I can observe that Requests is still the default (e.g. not optional as I though you would move toward).

Can I get guidance on how you want to proceed so that we can remain efficient?

On another subject, I quietly improved the "firewall blocker" situation.

```python
import niquests
import requests
import httpx

URL = "https://kick.com/video/f60e4115-7b7b-4680-a4b9-a48e7b74d45c"

if __name__ == "__main__":

    r = requests.get(URL)
    print("Requests", r)

    r = httpx.get(URL)
    print("HTTPX", r)

    r = niquests.get(URL)
    print("Niquests", r)
```

```
Requests <Response [403]>
HTTPX <Response [403 Forbidden]>
Niquests <Response HTTP/2 [200]>
```

Regards,

**Comment by Grub4K** [comment]
Currently `requests` is still in the dependencies, however, in the future we want to transition towards a `default` group so that you get no dependencies when installing `yt-dlp` but default ones when installing `yt-dlp[default]`. In the same vein as `curl_cffi`, `niquests` should then become a new group, which means it would be the default request handler when installing `yt-dlp[niquests]`.

As mentioned by pukkandan before:
> Separate requests and niquests into two files, creating any necessary abstraction to deal with common code.

I think merging common code can wait for now and happen later when code is finalized, but we should separate the two and provide both as an option through dependency groups. I can handle the related boilerplate of adding a new request handler and its group if you would like.

---
Side note, awesome to see that `niquests` seems resilient to the detection. If it provides a better default experience out of the box this could be huge for UX. Im excited for where this can go.

**Comment by Ousret** [comment]
OK! Noted.
I am probably going to open a new PR that will supersede this one.

Two last questions:

- So, if everything is optional, and let's say one user have all the optional net tools installed, how do you choose what to use?
- Happy Eyeballs support landed in [Niquests](https://github.com/jawah/niquests/releases/tag/v3.5.5), it will surely be of use, WDYT?

**Comment by Grub4K** [comment]
> So, if everything is optional, and let's say one user have all the optional net tools installed, how do you choose what to use

This is currently not implemented, but request handler preference will from cli will be implemented in the future.

> Happy Eyeballs support landed in [Niquests](https://github.com/jawah/niquests/releases/tag/v3.5.5), it will surely be of use, WDYT

Im not sure how it would interface with the core code. Theoreticalyl it could be helpful, but wed have to see where edge cases arise

**Comment by pukkandan** [comment]
> > So, if everything is optional, and let's say one user have all the optional net tools installed, how do you choose what to use
> 
> This is currently not implemented, but request handler preference will from cli will be implemented in the future.

For the PR, we can hardcode a priority > requests. It reasonable to assume for now that user will only install niquests if they want to use it.

**Comment by Ousret** [comment]
OK. I noted and applied all of your concerns.
See https://github.com/yt-dlp/yt-dlp/pull/9542 to look at the mergeable PR.

> That is not necessarily true. It should be depriotized by default and users can prioritize it if they want IMO. Requests/urllib must be the default regardless of what is installed, for now.

I strongly disagree, I think that it add an unnecessary friction for users. Due to current situation, requests is almost guaranteed to be there even after you decide to mark it optional. If you do as you said, I bet that almost no one will actually leverage other network handlers.

I am closing this PR as per #9542 

Regards,

**Comment by absidue** [comment]
@Ousret I get that niquests is your baby, but I agree with the yt-dlp devs, it should be opt-in for the moment. yt-dlp supports a lot of sites, some of them are quite weird, niquests needs to prove that it can handle all that weirdness and edge cases. The only way that is going to happen is if people have a chance to use it and push it to it's limits. Making it the default shouldn't even be considered until it has proven that it works well in the context of yt-dlp, because the requests handler is a core part of yt-dlp and if that breaks, you can't download anything.

**Comment by coletdjnz** [comment]
Don't worry about the preference atm, we will discuss this internally. I haven't had a chance to have a good look at things yet.

**Pull request closed without merging**